### PR TITLE
Add missing tests for the case where packages are referenced with PrivateAssets="All" attribute

### LINKED ISSUE #10397: Missing tests for the case where packages are referenced with PrivateAssets="All" attribute
## Details about Problem

NuGet version: dev

## Detailed repro steps so we can see the same problem

There are no `RemoteDependencyWalker` tests for the case where packages are referenced with PrivateAssets="All" attribute, e.g.:
```
    -> Lib1 -> PackageA [1.0.0] (PrivateAssets=All)
App
    -> Lib2 -> PackageA [2.0.0] (PrivateAssets=All)
```

Because of that these lines don't have any test coverage: https://github.com/NuGet/NuGet.Client/blob/b5b44526dea0379ebd6c8e51e8a041c06d5845ca/src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs#L166-L169

### PR DESCRIPTION
**Pull request by marcin-krystianc**
Adds missing tests for the dependency walker when packages are referenced with `PrivateAssets="All"`.
Fixes: https://github.com/NuGet/Home/issues/10397

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 1]
```diff
@@ -1682,7 +1828,7 @@ private void AssertPath<TItem>(GraphNode<TItem> node, params string[] items)
                 node = node.OuterNode;
             }
 
-            Assert.Equal(matches, items);
```
Comment (marcin-krystianc) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1685]: `Expected` was swapped with `actual`

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 2]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+                var downgradedBy = result.Downgrades[0].DowngradedTo;
+
+                AssertPath(downgraded, "A 1.0.0", "B 1.0.0", "C 1.0.0", "D 2.0.0");
+                AssertPath(downgradedBy, "A 1.0.0", "B 1.0.0", "D 1.0.0");
+            }
+        }
+
+        /// <summary>
+        /// A -> B 1.0.0 (PrivateAssets=?) -> C 2.0.0
+        ///   -> C 1.0.0
```
Comment (erdembayar) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1770]: Can you check accuracy of this graph?
`B` version is `2.0.0` not `1.0.0`
Also D is no where here.

File: test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs [hunk 3]
```diff
@@ -154,6 +154,7 @@ public TestPackage DependsOn(string id, string version, LibraryDependencyTarget
                     },
                     VersionCentrallyManaged = versionCentrallyManaged,
                     ReferenceType = libraryDependencyReferenceType,
+                    SuppressParent = privateAssets ?? LibraryIncludeFlagUtils.DefaultSuppressParent,
```
Comment (donnie-msft) [inline test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs]: Thinking we shouldn't set `SuppressParent` unless a privateAssets was specified. 
If we specify the default in the test, then `LibraryDependency` changes the default property value, the test won't be accurate, right?

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 4]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+
+            provider.Package("D", "1.0.0");
+            provider.Package("D", "2.0.0");
+
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
+            var result = rootNode.Analyze();
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit type here

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 5]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+
+            // Assert
+            Assert.Equal(expectedDowngrades, result.Downgrades.Count);
+            Assert.Empty(result.VersionConflicts);
+
+            if (expectedDowngrades == 1)
+            {
+                var downgraded = result.Downgrades[0].DowngradedFrom;
+                var downgradedBy = result.Downgrades[0].DowngradedTo;
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit type here

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 6]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+                .DependsOn("D", "2.0.0", privateAssets: privateAssets2);
+
+            provider.Package("D", "1.0.0");
+            provider.Package("D", "2.0.0");
+
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit type here

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 7]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+            var result = rootNode.Analyze();
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit type here

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 8]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+
+            // Assert
+            Assert.Equal(expectedDowngrades, result.Downgrades.Count);
+            Assert.Empty(result.VersionConflicts);
+
+            if (expectedDowngrades == 1)
+            {
+                var downgraded = result.Downgrades[0].DowngradedFrom;
+                var downgradedBy = result.Downgrades[0].DowngradedTo;
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit type here

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 9]
```diff
@@ -1669,7 +1669,153 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+
+            provider.Package("D", "1.0.0");
+            provider.Package("D", "2.0.0");
 
+            context.LocalLibraryProviders.Add(provider);
+            var walker = new RemoteDependencyWalker(context);
+
+            // Act
+            var rootNode = await DoWalkAsync(walker, "A", framework);
+            var result = rootNode.Analyze();
```
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: explicit types here

**Review by donnie-msft — COMMENTED** [review]
Thanks for helping us improve test coverage!

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 2 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1770]: Thanks for catching this 👍 I've also simplified the graph from the previous test.

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 9 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done.

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 6 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 7 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 8 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 4 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 5 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs]: Done

File: test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs (hunk 3 above)
Comment (marcin-krystianc) [reply test/TestUtilities/Test.Utility/DependencyResolver/DependencyProvider.cs]: I agree, I've also done the same for the `libraryDependencyReferenceType`.

**Review by donnie-msft — APPROVED** [review]

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs [hunk 10]
```diff
@@ -1669,7 +1669,144 @@ public async Task CentralTransitiveDependency_Downgrades_IfDowngrades_WereNotRej
...
+        [InlineData(LibraryIncludeFlags.All, LibraryIncludeFlags.All, 0)]
+        public async Task PrivateAssetsAll_VersionConflicts(LibraryIncludeFlags? privateAssets1,
+            LibraryIncludeFlags? privateAssets2, int expectedConflicts)
+        {
+            var framework = NuGetFramework.Parse("net45");
+            var context = new TestRemoteWalkContext();
+            var provider = new DependencyProvider();
+
+            provider.Package("A", "1.0.0")
+                .DependsOn("B", "1.0.0")
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1692]: This code allows it, but in practice B *has* to be a project.

**Review by nkolev92 — APPROVED** [review]
We non ideal coverage for some of these scenarios. 
We have lots of E2E/nuget.exe tests for these scenarios, so it's often hard to determine what's actually tested. Appreciate the help!

Thanks!

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 1 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1685]: I'm probably guilty of so many of those :D

File: test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs (hunk 10 above)
Comment (marcin-krystianc) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1692]: Yeah, I was thinking about that too, Unfortunately the `Test.Utility.DependencyProvider` doesn't let to mock  projects. On the other hand, projects and packages are handled equally by the graph walking algorithm so it doesn't really matter.

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.DependencyResolver.Core.Tests/RemoteDependencyWalkerTests.cs:1692]: Yep, it was mostly informational for any future readers. The test is going through the intended codepath.

**Comment by donnie-msft** [comment]
@marcin-krystianc There are some formatting errors:

>   test\TestUtilities\Test.Utility\DependencyResolver\DependencyProvider.cs(152,39): Fix whitespace formatting.
>   test\TestUtilities\Test.Utility\DependencyResolver\DependencyProvider.cs(152,83): Fix whitespace formatting.

Apologies that this wasn't well documented. It was added this year to our pipeline, but not to our docs. Here's my PR to add instructions: https://github.com/NuGet/NuGet.Client/pull/3817

**Comment by marcin-krystianc** [comment]
> @marcin-krystianc There are some formatting errors:
> 
> > test\TestUtilities\Test.Utility\DependencyResolver\DependencyProvider.cs(152,39): Fix whitespace formatting.
> > test\TestUtilities\Test.Utility\DependencyResolver\DependencyProvider.cs(152,83): Fix whitespace formatting.
> 
> Apologies that this wasn't well documented. It was added this year to our pipeline, but not to our docs. Here's my PR to add instructions: #3817

Ah I remembered to run the `dotnet format` for the initial PR, but then I forgot to run it again after making further changes. Fixed now.

**Pull request merged**